### ISSUE TITLE
Cubical: no canonicity for record types without η-equality

### ISSUE THREAD

**Issue by nad** [agda/agda#6581]
The following code is accepted:
```agda
{-# OPTIONS --safe --cubical #-}

open import Agda.Primitive.Cubical
open import Agda.Builtin.Equality
open import Agda.Builtin.Nat

record Σ (A : Set) (P : A → Set) : Set where
  no-eta-equality
  constructor _,_
  field
    fst : A
    snd : P fst

open Σ public

test : Σ Nat (λ _ → Nat)
test = primTransp (λ _ → Σ Nat (λ _ → Nat)) i0 (zero , zero)

_ : test .fst ≡ _
_ = refl

_ : test .snd ≡ _
_ = refl
```
However, the following code is not accepted:
```agda
_ : test ≡ (_ , _)
_ = refl
```
Here `test` is a closed term, so the last piece of code should also be accepted.

The following code is accepted, so the problem seems to be related to record types without η-equality:
```agda
{-# OPTIONS --safe --cubical #-}

open import Agda.Primitive.Cubical
open import Agda.Builtin.Equality
open import Agda.Builtin.Nat

data Σ (A : Set) (P : A → Set) : Set where
  _,_ : (x : A) → P x → Σ A P

test : Σ Nat (λ _ → Nat)
test = primTransp (λ _ → Σ Nat (λ _ → Nat)) i0 (zero , zero)

_ : test ≡ (_ , _)
_ = refl
```

**Comment by andreasabel** [comment]
> the problem seems to be related to record types without η-equality:

Presumably, they should be treated like single-constructor data types?

**Comment by plt-amy** [comment]
The implementation of transp for records (eta or not) is, internally, a definition by copatterns (an _ordinary_ definition!). Being an ordinary definition means we spend very little time outside the abstract machine to compute Kan operations (all we need is head-normalization of the _type_), and being defined by copatterns means that transport won't accidentally expand large copattern-matching definitions.

The implementation of `transp` as a copattern-matching definition also extends very smoothly to coinductive records, and transporting circular values.

**Comment by plt-amy** [comment]
To be clear, this behaviour can be surprising (so it should be documented), but it's not a bug, IMO.

**Comment by szumixie** [comment]
It has the same behavior if the record is marked with `pattern`, which genuinely breaks canonicity:
```agda
{-# OPTIONS --safe --cubical #-}

open import Agda.Primitive.Cubical
open import Agda.Builtin.Equality
open import Agda.Builtin.Nat

record Σ (A : Set) (P : A → Set) : Set where
  no-eta-equality
  pattern
  constructor _,_
  field
    fst : A
    snd : P fst

open Σ public

test : Σ Nat (λ _ → Nat)
test = primTransp (λ _ → Σ Nat (λ _ → Nat)) i0 (zero , zero)

f : Σ Nat (λ _ → Nat) → Nat
f (x , y) = x

_ : f test ≡ zero
_ = refl -- f (primTransp (λ _ → Σ Nat (λ _ → Nat)) i0 (zero , zero)) != zero
```

**Comment by plt-amy** [comment]
Yeah, I guess _that_ is a bug

**Comment by andreasabel** [comment]
Fixed by #6583.
